messagequeue/messagequeue.go

```diff
@@ -0,0 +1,216 @@
+	}
+
+	// TODO: add a msg.Combine(...) method
+	// otherwise, combine the one we are holding with the
+	// one passed in
+	for _, e := range entries {
+		if e.Cancel {
+			if mq.wl.Remove(e.Cid, ses) {
+				work = true
+				mq.out.Cancel(e.Cid)
+			}
+		} else {
+			if mq.wl.Add(e.Cid, e.Priority, ses) {
+				work = true
+				mq.out.AddEntry(e.Cid, e.Priority)
+			}
+		}
+	}
+}
+
+func (mq *msgQueue) Startup(ctx context.Context, initialEntries []*wantlist.Entry) {
+
+	// new peer, we will want to give them our full wantlist
+	fullwantlist := bsmsg.New(true)
+	for _, e := range initialEntries {
+		for k := range e.SesTrk {
+			mq.wl.AddEntry(e, k)
+		}
+		fullwantlist.AddEntry(e.Cid, e.Priority)
+	}
+	mq.out = fullwantlist
+	mq.work <- struct{}{}
+
+	go mq.runQueue(ctx)
+}
+
+func (mq *msgQueue) Shutdown() {
+	close(mq.done)
+}
+func (mq *msgQueue) runQueue(ctx context.Context) {
+	for {
+		select {
+		case <-mq.work: // there is work to be done
+			mq.doWork(ctx)
+		case <-mq.done:
+			if mq.sender != nil {
+				mq.sender.Close()
+			}
+			return
+		case <-ctx.Done():
+			if mq.sender != nil {
+				mq.sender.Reset()
+			}
+			return
+		}
+	}
+}
+
+func (mq *msgQueue) doWork(ctx context.Context) {
+	// grab outgoing message
+	mq.outlk.Lock()
+	wlm := mq.out
+	if wlm == nil || wlm.Empty() {
+		mq.outlk.Unlock()
+		return
+	}
+	mq.out = nil
+	mq.outlk.Unlock()
+
+	// NB: only open a stream if we actually have data to send
+	if mq.sender == nil {
+		err := mq.openSender(ctx)
+		if err != nil {
+			log.Infof("cant open message sender to peer %s: %s", mq.p, err)
+			// TODO: cant connect, what now?
+			return
+		}
+	}
+
+	// send wantlist updates
+	for { // try to send this message until we fail.
+		err := mq.sender.SendMsg(ctx, wlm)
+		if err == nil {
+			return
+		}
+
+		log.Infof("bitswap send error: %s", err)
+		mq.sender.Reset()
+		mq.sender = nil
+
+		select {
+		case <-mq.done:
+			return
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Millisecond * 100):
+			// wait 100ms in case disconnect notifications are still propogating
+			log.Warning("SendMsg errored but neither 'done' nor context.Done() were set")
+		}
+
+		err = mq.openSender(ctx)
+		if err != nil {
+			log.Infof("couldnt open sender again after SendMsg(%s) failed: %s", mq.p, err)
+			// TODO(why): what do we do now?
+			// I think the *right* answer is to probably put the message we're
+			// trying to send back, and then return to waiting for new work or
+			// a disconnect.
+			return
+		}
+
+		// TODO: Is this the same instance for the remote peer?
+		// If its not, we should resend our entire wantlist to them
+		/*
+			if mq.sender.InstanceID() != mq.lastSeenInstanceID {
+				wlm = mq.getFullWantlistMessage()
+			}
+		*/
+	}
+}
+
+func (mq *msgQueue) openSender(ctx context.Context) error {
+	// allow ten minutes for connections this includes looking them up in the
+	// dht dialing them, and handshaking
+	conctx, cancel := context.WithTimeout(ctx, time.Minute*10)
```

Comment:
The timeout value could potentially be a configuration field of the `msgQueue` struct? Not sure though